workspaces/spectacle/src/index.ts

```diff
@@ -110,10 +114,23 @@ export interface IOpticContext {
 
 ////////////////////////////////////////////////////////////////////////////////
 
+export type AsyncStatus<T> =
+  | {
+      loading: true;
+      error: false;
+      data: null;
+    }
+  | {
+      loading: false;
+      error: true;
+      data: null;
+    }
+  | { loading: false; error: false; data: T };
+
 export interface IBaseSpectacle {
-  query(options: SpectacleInput): Promise<any>;
+  query(options: SpectacleInput): Promise<AsyncStatus<any>>;
```

Comment:
We can change query + mutate to be generic and pass through their generic to AsyncStatus - will update types in a separate PR since I want to keep this PR (somewhat) focused and not digress too much

Reply:
Ok, this actually requires more work to update the typing (since some error keys have non-boolean values - i.e. error types) - reverting for now